plugins/inputs/t128_tank/reader.go

```diff
@@ -0,0 +1,368 @@
+	r.tankReadCmdCtx = tankReadCmdCtx
+	return r
+}
+
+func (r *Reader) Run(mainCtx context.Context) {
+	readCtx, readCtxCancel := context.WithCancel(mainCtx)
+	lastIndex, err := r.getIndex(r.indexPath, r.defaultIndex)
+	r.lastSavedIndex = lastIndex.value
+	if err != nil {
+		r.log.Errorf("Error in get index %v", err)
+		readCtxCancel()
+		return
+	}
+
+	nextSaveCheck := time.After(2 * time.Second)
+	defer func() {
+		readCtxCancel()
+		r.setIndex(r.indexPath, index{value: r.lastSavedIndex})
+	}()
+
+	go r.read(readCtx, lastIndex.next())
+
+	for {
+		select {
+		case <-mainCtx.Done():
+			r.log.Errorf("%s reader done", r.topic)
+			return
+		case <-nextSaveCheck:
+			if lastIndex.value > r.lastSavedIndex {
+				r.setIndex(r.indexPath, lastIndex)
+				r.lastSavedIndex = lastIndex.value
+			}
+			nextSaveCheck = time.After(2 * time.Second)
+		case err := <-r.readDone:
+			var errBoundaryFault *boundaryFault
+			if err != nil && errors.As(err, &errBoundaryFault) {
+				r.log.Debugf("detected boundary fault, restarting %s tank read from index %d", r.topic, errBoundaryFault.nextAvailableIndex.value)
+				go r.read(readCtx, errBoundaryFault.nextAvailableIndex)
+			} else {
+				go r.read(readCtx, lastIndex.next())
+			}
+		}
+	}
+}
+
+func (r *Reader) getIndex(indexPath string, defaultIndex index) (index, error) {
+	if indexPath == "" {
+		r.log.Debugf("index file path not provided, starting with default index %s", defaultIndex.string())
+		return defaultIndex, nil
+	}
+	content, err := os.ReadFile(indexPath)
+	if errors.Is(err, os.ErrNotExist) {
+		r.log.Debugf("index file %s does not exist, starting with default index %s", indexPath, defaultIndex.string())
+		return defaultIndex, nil
+
+	} else if err != nil {
+		return defaultIndex, fmt.Errorf("encountered error reading index file, starting with default index %s: %s", defaultIndex.string(), err)
+	}
+
+	r.log.Debugf("found '%s' in index file", content)
+
+	index, err := newIndex(string(content))
+	if err != nil {
+		return defaultIndex, fmt.Errorf("encountered error while parsing index file content, starting with default index %s: %s", defaultIndex.string(), err)
+	}
+
+	return index, nil
+}
+
+func (r *Reader) setIndex(indexPath string, index index) {
+	if indexPath == "" {
+		r.log.Debugf("index file path not provided, starting with index %d", index.value)
+		return
+	}
+	_, err := os.Stat(path.Dir(indexPath))
+	if errors.Is(err, os.ErrNotExist) {
+		err := os.MkdirAll(path.Dir(indexPath), os.ModePerm)
+		if err != nil {
+			r.log.Debugf("unable to create index file parent directory: %s", err)
+			return
+		}
+	}
+
+	err = os.WriteFile(indexPath, []byte(index.string()), 0600)
+	if err != nil {
+		r.log.Debugf("unable to update index to %d: %s", index, err)
+	}
+}
+
+func (r *Reader) read(readCtx context.Context, startingIndex index) {
+	r.log.Errorf("starting %s tank read from index %d", r.topic, startingIndex.value)
```

Comment:
This isn't an error log and will be hit a lot. I'd say this is a debug log

Reply:
good catch; maybe useful as an info log? I'm not sure